optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/variable/index/IndexVariableListener.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright 2021 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.core.impl.domain.variable.index;
+
+import java.util.List;
+import java.util.Objects;
+
+import org.optaplanner.core.api.domain.variable.VariableListener;
+import org.optaplanner.core.api.score.director.ScoreDirector;
+import org.optaplanner.core.impl.domain.variable.descriptor.ListVariableDescriptor;
+import org.optaplanner.core.impl.score.director.InnerScoreDirector;
+
+public class IndexVariableListener<Solution_> implements VariableListener<Solution_, Object>, IndexVariableSupply {
+
+    protected final IndexShadowVariableDescriptor<Solution_> indexShadowVariableDescriptor;
+    protected final ListVariableDescriptor<Solution_> sourceListVariableDescriptor;
+
+    public IndexVariableListener(
+            IndexShadowVariableDescriptor<Solution_> indexShadowVariableDescriptor,
+            ListVariableDescriptor<Solution_> sourceListVariableDescriptor) {
+        this.indexShadowVariableDescriptor = indexShadowVariableDescriptor;
+        this.sourceListVariableDescriptor = sourceListVariableDescriptor;
+    }
+
+    @Override
+    public void beforeEntityAdded(ScoreDirector<Solution_> scoreDirector, Object entity) {
+        // Do nothing
+    }
+
+    @Override
+    public void afterEntityAdded(ScoreDirector<Solution_> scoreDirector, Object entity) {
+        insert((InnerScoreDirector<Solution_, ?>) scoreDirector, entity);
+    }
+
+    @Override
+    public void beforeVariableChanged(ScoreDirector<Solution_> scoreDirector, Object entity) {
+        // TODO maybe set all indexes to null
+        // When we support over-constrained planning (e.g. there is not enough Employees to assign all tasks), then
+        // an unassigned planning value should have a null index.
```

Comment:
Yes, we need tasks that go from assigned to unassigned (for example CH undo move) to set their index to null, to calculate to the score correct (for example CH undo move full assert).

Reply:
This is a Phase 1 implementation (= mergable to main).
It is correct, not optimization fast wise yet, but that is for phase 2.


A1 {Ann index 7 -> Beth index 5}

when moving A1
  before(Ann) => Ann.taskList elements has all index=null. => before(A0, A1, A2, A3) + delayed after (A0, A1, A2, A3)
  after(Ann) delayed
  before(Beth) => Beth.taskList elements has all index=null. => before(B0, B1) + delayed after (B0, B1)
  after(Beth) delayed
  after(Ann) => Ann.taskList elements get index 0, 1, 2, ... => before(A0, A2, A3) + delayed after (A0, A2, A3)
  after(Beth) => Beth.taskList elements get index 0, 1, 2, ... => before(B0, A1, B1) + delayed after (B0, A1, B1)
  

when swapping A1 with A2
  before(Ann) + delayed after(Ann)
  before(Beth) + delayed after(Beth)

Reply:
Will be addressed together with CH (https://issues.redhat.com/browse/PLANNER-2302) or overconstrained planning.

Reply:
By the way, doing this non-incrementally, i.e. setting all index and inverse variables to `null` before variable changes is a performance killer. I can see -75% score calculation speed on the largest data set (500 tasks, 20 employees).